Add plenum zones to Room2D.to_honeybee() method

saeranv opened the pull request:
@chriswmackey, cc @mostaphaRoudsari 

This is still WIP, here mainly for reference at this point. 

Currently this method extrudes ceiling plenums (as a Honeybee Room, assuming Rooms are equivalent to thermal zones) to a height based on a user-provided input, and intersects the plenum walls with adjacent rooms if they are at different heights. All boundary conditions adiabatic except for plenum walls exposed to outdoors, or are adjacent to the ceiling of the underlying thermal zone. 

The one exception to this is the ceiling of the plenum zone is currently set to 'Outdoors' - when it should be the Surface of the zone on top, (which is why one of my unit tests are failing). This is an expected error at this point since I don't know how to get the vertically adjacent face information to determine height extrusion/adjacency information. 

I posted a couple of questions here: https://github.com/ladybug-tools/dragonfly-core/issues/20 that will help me understand how to finish this up. 

Thanks

chriswmackey commented on dragonfly/room2d.py:
```diff
@@ -756,7 +756,7 @@ def remove_colinear_vertices(self, tolerance=0.01):
         return rebuilt_room
 
-    def to_honeybee(self, multiplier=1, tolerance=0.01):
+    def to_honeybee(self, multiplier=1, tolerance=0.01, add_plenums=False, plenum_ht=1.0):
```
Also, this is a minor styling comments but, for input arguments, I usually like to write out the full word (eg. `plenum_height`) instead of using abbreviations. Inside the method, I use abbreviations a lot but the inputs are going to be read by more than just the source code developers so I feel it's better to be explicit.

saeranv replied:
@chriswmackey what's the best way to distinguish or combine floor and ceiling plenums. Should I add a `add_floor_plenum` AND a `add_ceiling_plenum` argument to the method, or should I just assume there's always a floor and ceiling plenum. 

I think it makes sense to have just a ceiling plenum set as default, and having the floor plenum as an option (from a modeler's perspective) since that is the most common plenum scenario. So something like: 
```
to_honeybee(self, .... add_ceiling_plenum=True, add_floor_plenum=False)
```

?

saeranv replied:
@chriswmackey 

As I'm implementing the floor_plenum, I realize I don't know what is the best rule to autocalculate floor plenum height. Should it be 1/2 the distance we calculate from parent.floor_to_floor - self.floor_to_ceiling? This would mean that when floor plenums are set, we are assuming there's a ceiling plenum that's filling the rest of the gap between vertically adjacent rooms. 

Alternatively it can be the full height of the story/room gap (like how we have the ceiling plenum height right now) - but that means that we can only have a floor OR ceiling plenum, but never both. 

What's the desired behaviour?

chriswmackey commented on dragonfly/room2d.py:
```diff
@@ -769,7 +769,10 @@ def to_honeybee(self, multiplier=1, tolerance=0.01):
                 This is also used in the generation of Windows. Default: 0.01,
                 suitable for objects in meters.
+            add_plenums: Boolean to auto-generate floor and ceiling plenums for the Room
+                (Default: False).
```
Missing docstring for plenum height.

saeranv replied:
I'll remove the plenum_height argument and set it from parent's floor to floor.

chriswmackey commented on dragonfly/room2d.py:
```diff
@@ -1061,6 +1134,90 @@ def _flip_wall_assigned_objects(original_geo, bcs, win_pars, shd_pars):
+                new_faces.append(face)
+            else:
+                adj_rm = self._parent.room_by_identifier(
+                    bc.boundary_condition_objects[-1])
```
I see that you already need for the Room2D to have a parent Story in order for this code to work here. So, in light of this, I would suggest getting rid of the `plenum_ht` input and just raise an exception if `add_plenums` is True and the Room2D has no parent Story.

chriswmackey replied:
Also, in the case that the Room2D's ceiling height aligns with the top of the story to within the tolerance, no plenum should be generated.

chriswmackey commented on dragonfly/room2d.py:
```diff
@@ -1016,6 +1033,62 @@ def intersect_adjacency(room_2ds, tolerance=0.01):
+                plenum_hb_room[i + 1].boundary_condition = bc
+            # else:
+            #     adjacencies.append(
+            #         (plenum_hb_room[i + 1], bc.boundary_condition_objects))
```
If I am being honest, I was originally thinking that this situation of a Surface boundary condition for a plenum wall would be simpler (as in, we just make it Adiabatic).  This is because most plenums would be adjacent to other plenums and so you would have to figure out what this adjacent plenum's ID is (I guess that's not too difficult if we are using a standard naming convention). I guess if you have it mostly figured out, we can go that route.

chriswmackey replied:
Maybe to put it another way, I would much rather get this method working correctly for both floor and ceiling plenums than to have it correctly doing this splitting and adjacency setting correctly.  And it just feels like this splitting could be a rabbit hole right now that is going to stop us from reaching this first goal. So I would ask: do you feel that you have this splitting figured out enough to reliably get it working for both floor and ceiling plenums?

saeranv replied:
I agree with you, I think it's better to get the simpler scenario working first and then build on complexity from there. Additionally, if I'm interpreting your comments correctly, and the plenum walls are going to be adiabatic anyway, there isn't a benefit to splitting plenum walls.

chriswmackey commented on dragonfly/room2d.py:
```diff
@@ -1061,6 +1134,90 @@ def _flip_wall_assigned_objects(original_geo, bcs, win_pars, shd_pars):
+                    # No need to split the surface along its height
+                    new_faces.append(face)
+                elif flr_diff > tolerance and ciel_diff > tolerance:
+                    # split the face into to 3 smaller faces along its height
```
As I understand it, this case really shouldn't be happening for a plenum (a plenum would only ever be split in two and not three). Or I guess you have this case because you exposed `plenum_ht` as an input. Let's get rid of that input and then we can always assume that all plenums are going to the same height.

saeranv replied:
As discussed above, I'll get rid of this for now and build up the more important plenum features first (i.e getting BCs and geometry working).

saeranv commented:
@chriswmackey 

This update (should) finish up everything for the ceiling plenum. I removed the adjacent face splitting, got the autocalculation of plenum height from the parent Story, updated all the docstrings, and set the BCs according to what we discussed. 

The final thing remaining is how to add floor plenums into this method. I left some questions to your comments above which provide more detail.

chriswmackey commented:
Thanks @saeranv .

The adiabatic boundary condition type is specific to honeybee-energy since we decided that it does not have any meaning for daylight simulation. Furthermore, dragonfly-core doesn't depend upon honeybee-energy (only honeybee-core). So, to handle this situation, you should try to get the adiabatic boundary condition [using a try/except loop like you see here](https://github.com/ladybug-tools/dragonfly-core/blob/master/dragonfly/room2d.py#L826-L830). If there's no adiabatic boundary condition available, you can fall back on the outdoor boundary condition, which is fine for outdoor studies.

For auto-calculating the floor plenum height, you should find the difference between the [Room2D floor_height](https://github.com/ladybug-tools/dragonfly-core/blob/master/dragonfly/room2d.py#L471-L473) and the parent [Story floor_height](https://github.com/ladybug-tools/dragonfly-core/blob/master/dragonfly/story.py#L180-L186). If this difference is less than the tolerance, than no floor plenum should be added.  The vast majority of time, there will be no difference between these two and so there will be no floor plenums. So I think you don't really need separate inputs on `Room2D.to_honeybee` to distinguish the floor plenum from the ceiling plenum.

Also FYI, I will eventually add support for people to hard-assign a floor height to a Story that overrides the one that's currently derived from the rooms. So, in those cases, the floor plenums will get used a lot more.

saeranv commented:
@chriswmackey 

I've taken this PR out of draft status, the method should be done now, and ready for final review. I made two changes in this latest PR: I implemented the floor plenums, and also consolidated any plenum calculations in the to_honeybee method to it's own function. I also added a new test to test floors, and floors and plenums. 

Also note that I decided to output the result of the to_honeybee method (with plenums) as a list of the honeybee Rooms. Let me know if you have a different way in mind to output the results of this add_plenum method. 

This is an example of how it works:
![image](https://user-images.githubusercontent.com/1685956/85910389-7ae1ba80-b7ec-11ea-85d8-cd4d3ebef817.png)

chriswmackey commented:
Thanks @saeranv . I will review shortly.

chriswmackey commented:
Ah crud. I was just merging a bunch of PRs and I accidentally merged this. Well, it seems this was 95% of the way there and it's probably easier at this point for me to just make any tweaks myself that I would have suggested in review.  So I think we can just call this task done. Thanks again for adding this @saeranv !
